### ISSUE TITLE
Should we start a Discord?

### ISSUE THREAD

**Issue by Rougemoot** [kmonad/kmonad#242]
First of all: I don't think it should replace the discussion in the github issues, and I don't expect there would be much traffic at all in there, but:

I think some discussion on this board would benefit from the "instant-ness" of a Discord chat. Some quick support or easier troubleshooting would be nice.

Of course I expect there would be frenquent "file an issue on github", because this is great for keeping track of issues, much better than a Discord server would.

Thoughts?

**Comment by david-janssen** [comment]
I sort of enjoy being able to take distance, so I probably won't be hanging out. However, I'd be happy to hop on at a scheduled time to discuss certain issues in real-time if people'd thought they'd benefit from that.

So: from my POV: feel free, I might make use, but probably won't be a constant regular :)

**Comment by slotThe** [comment]
On Thu, May 13 2021 09:54, Rougemoot wrote:
> Thoughts?

Real-time sounds great!

Personally, I would advocate for an IRC channel (freenode hosts a lot of
free software projects) or a matrix room (these two could be bridged
very easily as well).  Being yet another project that only has
proprietary communication channels would make me a bit sad.

**Comment by vherrmann** [comment]
Till now I didn't contribute anything to kmonad, but I would feel especially sad about the project being hosted on discord and not a free messenger. I feel like that, since kmonad is the project, that made me learn haskell (i'm still on it), so that i'm able to contribute to it.

**Comment by thoelze1** [comment]
I would be interested in live chat too. I second @slotThe 's proposal of IRC

**Comment by Rougemoot** [comment]
Discord has a lot of great tools that makes it much more 2021 than IRC, so I'd much prefer that, but if people prefer IRC I guess that'd at least be something.

**Comment by joshskidmore** [comment]
+1 for Discord. Most tech projects I'm a part of these days are using Discord. Hell, I don't even have IRC these days.

**Comment by premell** [comment]
+1 for discord. Kmonad can be a bit difficult to get into so would be nice to have a place to ask questions and trubbleshoot having to open a new issue

**Comment by MaxGyver83** [comment]
I have never missed a chat for KMonad. But I also think that not every beginner's question should be labeled an issue. When I had a question about surfingkeys and created an issue, they told me that there is a [GitHub Discussions](https://github.com/brookhong/Surfingkeys/discussions) page for questions like that. I think not many GitHub repositories use it. But maybe this is an option for KMonad, too. (Of course, this could be used additionally to IRC or Discord.)

**Comment by premell** [comment]
I dont really care if its a github discussion forum or discord, but I think some way of communicating outside of creating issues would be nice

**Comment by david-janssen** [comment]
Maybe we should disappoint everyone and see if we can do some kind of ICQ room? Or am I showing my age with that joke?

I really like open-source free. Didn't realize discord wasn't that. However, I also really like ease of use and accessibility, and I remember having issues getting into IRC myself.

I'd really like to leave it open to the community, but on the other hand, a community without some guidance tends to go in 5 different directions, and suffer because of it. Sooo.....

How about we start a discord page. Basically because I think when a young nerd shows up at KMonad, they are probably more likely to know how to access discord than IRC, and some might be scared off by IRC. Whereas the old nerds are wiser and more patient, and although they prefer IRC, they are content using Discord and only occasionally making snide comments about the good old days?

However, anything worth preserving should never be 'stored on discord'. We should be ready to abandon ship the moment the policies start heading in the wrong direction. At which point the young nerds must adopt the ways of the wise ones, and off to IRC we go.

**Comment by david-janssen** [comment]
Oh, and if some of the wise ones want to already start an IRC to grumble about those little-enders in Discord, I'm not going to stop them, but I do think most of the signposts pointing at real-time help should point at Discord.

**Comment by vherrmann** [comment]
> I really like open-source free. Didn't realize discord wasn't that. However, I also really like ease of use and accessibility, and I remember having issues getting into IRC myself.

That it's not accessible isn't exactly true, since there is e.g. [kiwi](https://kiwiirc.com/nextclient/irc.freenode.net) that allows joining irc channels from the browser.

Also, matrix has with element a web client (you don't need your own server, since there is a public instance at matrix.org).

**Comment by david-janssen** [comment]
> That it's not accessible isn't exactly true, since there is e.g. kiwi that allows joining irc channels from the browser.

Hey, that's really cool! Much nicer than the last time I tried to get on IRC.

> Also, matrix has with element a web client (you don't need your own server, since there is a public instance at matrix.org).

I'm not sure I understand what you mean.

Since Kiwi gives *such* straightforward access to chat, my preference actually goes to IRC then. Any benefits to Discord that IRC doesn't have? Remember: you are talking to someone who has used neither: a mountain hermit who knows how to code, and very little else. That I can even maneuvre git is a miracle.

**Comment by david-janssen** [comment]
Also, I think we should wait at least until Monday to make any decisions, to give everyone a chance to contribute their views.

**Comment by vherrmann** [comment]
> I'm not sure I understand what you mean.

There is a web client available for the matrix protocol, called [element](https://element.io/get-started).

**Comment by david-janssen** [comment]
The matrix protocol... 

I found this, but it's confused me even more: https://www.matrix.com/

Is this a 3d option? IRC, Discord, or Matrix? Is it something like Stack?

**Comment by vherrmann** [comment]
Element is a client for matrix. Matrix is similar to irc, since it is an open protocol, but in contrast it's federated. You can pretty easily bridge irc and matrix, so i don't really care which if either of those is chosen. 

Matrix feels much more 2021 than irc, since it allows reactions, real replies, message editing, you don't have to be online all the time…

**Comment by vherrmann** [comment]
> Any benefits to Discord that IRC doesn't have?

Discord allows reactions, message edits, built-in replies (seeeolvable by irc clients), you can see who is online, you can see how is typing, you don't have to be online, you can have multiple channels, there is voice-chat and video-chat and screen capture

**Comment by vherrmann** [comment]
But you have to use the official client that isn't exactly that privacy-aware (There are alternative clients, but i've heard you get banned, the moment discords finds out you used one).

**Comment by david-janssen** [comment]
I'm thinking that KMonad might at some point have some success with people who like to game a lot and want to bind macros and such to complicated keymaps. I'm probably right in estimating they'd be much happier popping into Discord, right?

**Comment by vherrmann** [comment]
There are also some other issues, e.g. if there is a new version of discord available, you **have** to use it. Which is really annoying on linux, since they don't warn you.

**Comment by david-janssen** [comment]
Ahh, I've only ever used discord from my browser, and then only a few times. I really would like to keep things as easy as possible on Linux, since um... I use Linux.

**Comment by vherrmann** [comment]
It's not a real issue, but it just chose discords attitude

**Comment by david-janssen** [comment]
I'm going to actually write some code now and wait for opinions and views to trickle in. Amazing, how distracting social media can be :-D

**Comment by vherrmann** [comment]
>  I'm probably right in estimating they'd be much happier popping into Discord, right?
Yes, that's probably right.

 On the other hand, it wouldn't hurt to use irc, so that those that need reactions, editing, permanent messages, etc. can have that with a small effort on matrix and people, who have short questions can just hop onto kiwi.

**Comment by slotThe** [comment]
> Is this a 3d option? IRC, Discord, or Matrix? Is it something like
> Stack?

Essentially, but Matrix and IRC integrate with each other very easily
(the matrix project bridges all of freenode, for example, so we'd
essentially get that for free).

IRC webchat is indeed very easy to join (and probably poses the least
number hurdles for users who just want to ask a quick question); just
enter a username and the channel and be ready to go.

> Whereas the old nerds are wiser and more patient, and although they
> prefer IRC, they are content using Discord and only occasionally
> making snide comments about the good old days?

I can only speak for myself, but I won't be using discord anytime soon
(not that this should discourage the community from making one, should
more people want that).

> I'd really like to leave it open to the community, but on the other
> hand, a community without some guidance tends to go in 5 different
> directions, and suffer because of it. Sooo.....

I definitely think that a little BDFL action is necessary, should we not
be able to find a consensus :)

**Comment by Rougemoot** [comment]
There are also convenient ways to link a Discord server to IRC chatrooms, so that people can use whichever they like.

To me, Discord has a few advantages:

- People of 2021 are much more likely to look for a discord server than an IRC chatroom. Even in the FOSS community. Discord is currently the most used platform for such things, and for good reason imo.
- People might be turned off by an IRC chatroom, because let's face it, it's ancient, and not very user-friendly (just like with kiwi, on Discord you can join as a guest without creating an account IIRC.
- Multiple channels on a single server. It might seem trivial, but much more than a solution to "too much traffic to have a discussion" (which is a problem we most likely wouldn't have), it's a very nice way to compartimentalize discussion about different topics. Tech support, suggestions, general discussion, off-topic...)
- A very nice search engine, and discussion history. While I completely agree that nothing of "value" should exist only on discord (anything like bug reports, etc should *definitely* be kept on github, if only as a way to store them after having discussed them on Discord)

While I somewhat agree that using FOSS solutions is preferrable, I don't think it's worth giving ourselves any extra hurdles for the sake of doing so. That said, I'm definitely not a big contributor to the project, just a user who likes to give input and help where they can, and comes here for some help at times (although isn't that exatly the target for such a chatroom?)

**Comment by elkowar** [comment]
I'd **STRONGLY** suggest not going for IRC. Multliple reasons here, among them that IRC is definitely not all that friendly to younger people and not all that welcoming to beginners in general (as it doesn't allow for clean organization into different channels). but, most importantly: having messages not persist would be a _very_ stupid decision for any platform that's supposed to be focused around a tool like this, where a lot of the discussions _would_ also be people asking for help. IRC would inevitably end up with getting the same 3 questions every day, because it's literally impossible to look at history.

Matrix doesn't suffer from that last issue, and is thus already infinitely preferable to IRC.
**BUT**: Still strongly argue in favour of a Discord. Not only is discord by far the most welcoming platform among the bunch, and generally the most popular - but organization into separate channels is a _big_ thing for a project-focused community. even just splitting up development-related and user-focused discussions is a big deal. Not having that _immediately_ makes the place much less useful in general.

Additionally, it is still possible to set up IRC/matrix bridges for a discord server. This would have the big benefit of keeping the IRC fans (mostly) happy while still not losing out on the (imo very essential) benefits of discord (-> message history, search, etc).

Not going with discord would be purely idealistic choice, which,.... I'm not a fan of, especially when it comes to something as inherently _pragmatic_ as KMonad.

**Comment by slotThe** [comment]
On Fri, May 14 2021 10:47, ElKowar wrote:
>  but, most importantly: having messages not persist would be a _very_
>  stupid decision for any platform that's supposed to be focused around
>  a tool like this, where a lot of the discussions _would_ also be
>  people asking for help. IRC would inevitably end up with getting the
>  same 3 questions every day, because it's literally impossible to look
>  at history.

I don't think that that's very fair.  It's very easy to set up global
channel logging on IRC; both #haskell and #xmonad do it, for example.
This could then be placed inside the topic for everyone to easily find.

> Not only is discord by far the most welcoming platform among the
> bunch,

What could be more welcoming than IRC?  Just hop on and start chatting!
Don't even have to make a pseudo-account that you get pressured into
"claiming" later on.

> and generally the most popular - but organization into separate
> channels is a _big_ thing for a project-focused community. even just
> splitting up development-related and user-focused discussions is a big
> deal. Not having that _immediately_ makes the place much less useful
> in general.

Serious discussions surrounding development should probably happen in
issues/pr's anyways.  But really it's not that big of a deal.  Looking
at #xmonad again I think this illustrates quite well that things works
out just fine in general.  It's not like you can't have multiple
discussions on IRC at the same time either (highlights are generally
used in that case).

**Comment by david-janssen** [comment]
Ahhh, so there were strong opinions in both directions. This I did not
anticipate. I've given it some thought, and I said I would wait till Sunday, but
I've ended up landing on a both-a-and-b decision which I think will equally
please and disappoint everyone.

This should probably be turned into some documentation at some point, but for
now this seems good enough.

## TL;DR
Let's do both. I won't hang out in either. I'm happy to show up at scheduled
times or for appointments in both. I'm not going to be actively moderating
either. Anything that should last should be put on GitHub. Please be kind to 
each other. If you have serious issues with each other, come to me.

## My reasoning
### A or B
There are people who strongly prefer A, there are people who strongly prefer B.
Both have very valid and good reasons. Sometimes it is good for a BDfaW
(Benevolent Dictator for a While) to be very clear and forceful in their
decisions, but sometimes it is not.

I do not like disappointing people, or forcing my ways on people. However, when
it comes to the code-base, I am going to make the effort to impose my will when
I think it benefits the project. The discussion IRC-vs-Discord is more of an
argument along the lines of: do we want to socialize in the cafe next to the
library, or the arcade across the street? Honestly: I don't like cafes or
arcades, and I'll mostly stay in the library (or on top of the mountain behind
the library in solitude). Here of course the library means GitHub, which is
where I think we should keep a record of all important conversations.

So, you all have my blessing to start a Discord or an IRC (or an actual cafe or
arcade) and get as much cross-platform connectivity as you can manage. The
benefit of either is that we can have easier real-time interactions, and I'd be
happy to come and visit either for that purpose. I'm also happy to schedule a
weekly David-is-here-for-an-hour arrangement, or to pop in uninvited. However, I
am also one of those people who keeps their phone on airplane mode for 6 hours a
day, because I prefer to not to have my thought interrupted. I think it's
conducive to sanity.

If you ever notice that someone shows up and they have a question, and your
impulse is to refer them to an already existing record on either the IRC or the
Discord, then I would ask you to copy out that conversation, put it in a
well-titled issue on the github, and just close the issue, and refer people
there. You don't even have to edit it, if the conversation is clear, that's
enough. That way people will not have to enter a cafe or an arcade to access the
books. The books should live in the library.

### Moderation
If we are going to start hanging out in cafes or arcades, there will eventually
arise conflict. Such is the way of communities. Especially as they grow,
reputations get established, and egos get involved. Additionally, I am giving
the green-light on the opening of KMonad related real-time communities that I
won't be very real-time present in.

Everyone up to now has been extraordinarily pleasant to interact with, and it's
been an absolute pleasure to see my little booping-about project turn into
something that people like, want to use, and want to contribute to. As projects
grow, inevitably community moderation becomes a topic.

I am very much not a fan of stringent speech codes. I've seen communities get
very heated and upset around these topics, so I'm hoping to nip this in the bud
early. We will not be implementing one. I have a personal code that I live by
and use to guide my interactions (also outside of GitHub) that I am going to put
down here as an invitation. But again, this is just in anticipation of future
conflict. Up until now everyone has been absolutely great, so I am not asking
anyone to change anything about their behavior.

My code:
- Try your hardest to be kind
- Failing that, try your hardest to be civil
- Failing that, try your hardest to be quiet
- Failing that, go away until you can be quiet

If strong conflicts do arise, they can be brought to me, and I will hear out
both sides and attempt mediation. However, very much on a case-by-case basis.

### Conclusion
I hope I've given a clear overview of my thoughts, and that you can all find
yourselves in the outcome. Furthermore, I hope that the expectation I've
expressed about your behavior doesn't feel stifling. Having gotten all of this
out of the way, I will conclude with:

Go forth and socialize!

P.S. Closing because 'decision made', but I'm happy to hear feedback here.

**Comment by elkowar** [comment]
I like that decision! I'm assuming that @Rougemoot may be the one to create a discord server here, and @slotThe to manage an IRC?
I definitely think setting up bridging and having both is the way to go. That way, if there is a clear favorite, it will show itself to be that naturally. If there isn't, connecting the two places makes everything easy!

**Comment by david-janssen** [comment]
I will leave the decisions on who runs what up to the community. I will sit in the library and occasionally stroll through all available venues :-). Hey, who knows, maybe one day we'll have competing IRC channels and we can do dance-offs and rap-battles and such. I've been led to believe by my television that this is a thing that happens ;)

**Comment by elkowar** [comment]
Sounds good, i'll organize a beatboxer for the rapbattle!

**Comment by slotThe** [comment]
On Sat, May 15 2021 00:30, ElKowar wrote:
> and @slotThe to manage an IRC?

Sounds good to me :+1:.  I've contacted freenode staff to register the
kmonad project with the network, so we can get #kmonad, as well as the
whole #kmonad-* namespace (so competing channels can be called
#kmonad-join-us-were-much-better-than-them)!

**Comment by vherrmann** [comment]
*Meanwhile vherrmann already created the #kmonad channel on freenode*
:S

**Comment by elkowar** [comment]
oh lol

**Comment by slotThe** [comment]
> _Meanwhile vherrmann already created the #kmonad channel on freenode_
> :S

On freenode, [primary channels](https://freenode.net/kb/answer/namespaces) (one starting with a single `#`) are reserved for registered projects.  If you want to create a "community channel" of sort you'll have to use `##kmonad`.  But really, don't worry about it :)    Worst that'll happen is that freenode staff kicks you (and you lose your OP) and you just rejoin the channel

**Comment by premell** [comment]
Is there a reason  for not using github discussion by the way?

**Comment by david-janssen** [comment]
> Is there a reason for not using github discussion by the way?

Not really, never used it, never felt the need. What is the niche it could fill?

**Comment by Rougemoot** [comment]
Well, here we go!

https://discord.gg/3tFfWmnahN

Big thanks to @elkowar and @6gk for their help, they basically set the whole thing up. There's already a few people from the /r/unixporn discord server.

There are some things to do still, but we have a great start

**Comment by slotThe** [comment]
In case people want to stop by, we're currently squatting on `freenode/#kmonad` (join via [webchat](https://webchat.freenode.net/) if you don't have an IRC client set up!).  This will be official as soon as I hear back from the freenode admins (and perhaps once David sends them an email to confirm that I'm not just some rogue contributor acting on their own behalf :))

**Comment by 6gk** [comment]
@slotThe Nice.  Hope the response is timely!

What would your thoughts on bridging the IRC to the Discord?

I'll look into methods of doing that later, hopefully.

**Comment by zblesk** [comment]
Hi. How did this end? I can't connect to that channel. Neither the linked web client, nor my HexChat work. Is there currently a place for noob questions?

**Comment by slotThe** [comment]
> Hi. How did this end? I can't connect to that channel. Neither the linked web client, nor my HexChat work. Is there currently a place for noob questions?

Freenode imploded and we moved to Libera (as should be written in the current README): [#kmonad@libera.chat](https://web.libera.chat/#kmonad)
